src/main/java/build/buildfarm/worker/Worker.java

```diff
@@ -205,9 +207,20 @@ private boolean linkInputs(Path execDir, Digest inputRoot, Map<Digest, Directory
     for (FileNode fileNode : directory.getFilesList()) {
       Digest digest = fileNode.getDigest();
       Path fileCachePath = cacheDir.resolve(String.format("%s_%d", digest.getHash(), digest.getSizeBytes()));
-      OutputStream outputStream = Files.newOutputStream(fileCachePath);
-      instance.getBlob(digest).writeTo(outputStream);
-      outputStream.close();
+      boolean createFile = !Files.exists(fileCachePath);
+      if (createFile) {
+        OutputStream outputStream = Files.newOutputStream(fileCachePath);
+        instance.getBlob(digest).writeTo(outputStream);
+        outputStream.close();
+      }
+      if (createFile || Files.isExecutable(fileCachePath) != fileNode.getIsExecutable()) {
+        ImmutableSet.Builder<PosixFilePermission> perms = new ImmutableSet.Builder<PosixFilePermission>()
+          .add(PosixFilePermission.OWNER_READ);
+        if (fileNode.getIsExecutable()) {
```

Comment:
I decided to just modify the filename to correspond to the executable bit.  We'll double copy if the content matches, but I don't want to reimplement a filesystem to decouple file metadata from content.